File: typescript/src/advertisements/advertisement/infrastructure/stream/producer/events/AdvertisementApprovedEvent.ts
import {AdvertisementWasApproved} from "../../../../domain/events/AdvertisementWasApproved";
import {SerializableEvent} from "../../../../../../common/infrastructure/stream/SerializableEvent";

export class AdvertisementApprovedEvent extends SerializableEvent {
    private static readonly SCHEMA: string = 'https://demo.com/schemas/advertisement-approved_1_0.json';
    private static readonly SOURCE: string = 'advertisements';

    private constructor(
        id: string,
        schema: string,
        eventType: string,
        version: string,
        occurredOn: Date,
        correlationId: string,
        causationId: string | null,
        source: string,
        aggregateId: string,
        aggregateType: string,
        tenantId: string,
        payload: object
    ) {
        super(id, schema, eventType, version, occurredOn, correlationId, causationId, source, aggregateId, aggregateType, tenantId, payload);
    }

    public static create(advertisementWasApproved: AdvertisementWasApproved, correlationId: string, causationId: string | null, tenantId: string): AdvertisementApprovedEvent {
        const payload = {
            advertisementId: advertisementWasApproved.advertisementId,
        };

        return new AdvertisementApprovedEvent(
            advertisementWasApproved.id,
            AdvertisementApprovedEvent.SCHEMA,
            advertisementWasApproved.eventType,
            advertisementWasApproved.version,
            advertisementWasApproved.occurredOn,
            correlationId,
            causationId,
            this.SOURCE,
            advertisementWasApproved.aggregateId,
            advertisementWasApproved.aggregateType,
            tenantId,
            payload,
        );
    }
}

File: kotlin/src/main/kotlin/advertisements/advertisement/infrastructure/stream/producer/events/AdvertisementApprovedEvent.kt
package advertisements.advertisement.infrastructure.stream.producer.events

import advertisements.advertisement.domain.events.AdvertisementWasApproved
import common.infrastructure.stream.producer.SerializableEvent
import java.time.LocalDateTime

class AdvertisementApprovedEvent private constructor(
    override val id: String,
    override val schema: String,
    override val eventType: String,
    override val version: String,
    override val occurredOn: LocalDateTime,
    override val correlationId: String,
    override val causationId: String?,
    override val source: String,
    override val aggregateId: String,
    override val aggregateType: String,
    override val tenantId: String,
    override val payload: Map<String, Any>
) : SerializableEvent(id, schema, eventType, version, occurredOn, correlationId, causationId, source, aggregateId, aggregateType, tenantId, payload) {

    companion object {
        const val SCHEMA = "https://demo.com/schemas/advertisement-approved_1_0.json"
        const val SOURCE = "advertisements"

        fun create(advertisementWasApproved: AdvertisementWasApproved, correlationId: String, causationId: String?, tenantId: String): AdvertisementApprovedEvent {
            val payload = mapOf(
                "advertisementId" to advertisementWasApproved.advertisementId
            )

            return AdvertisementApprovedEvent(
                advertisementWasApproved.eventId,
                SCHEMA,
                advertisementWasApproved.eventType,
                advertisementWasApproved.version,
                advertisementWasApproved.occurredOn,
                correlationId,
                causationId,
                SOURCE,
                advertisementWasApproved.aggregateId,
                advertisementWasApproved.aggregateType,
                tenantId,
                payload,
            )
        }
    }
}
